### PR TITLE
Twitter get particular users stream

### PR DESCRIPTION
**Pull request by MoniMoledo**
Added to UsersTimelineDriver user stream.

Related to issue #65 

@JavierJia @aishwaryakapse

### CODE REVIEW

File: noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/TwitterFeedUsersTimelineDriver.java [hunk 1]
```diff
@@ -1,31 +1,43 @@
...
         ConfigurationBuilder builder = new ConfigurationBuilder();
-        builder.setDebugEnabled(true)
+        builder.setDebugEnabled(false)
                 .setOAuthConsumerKey(config.getConsumerKey())
                 .setOAuthConsumerSecret(config.getConsumerSecret())
                 .setOAuthAccessToken(config.getToken())
                 .setOAuthAccessTokenSecret(config.getTokenSecret())
                 .setJSONStoreEnabled(true);
-
+        //Get historical user data
```
Comment (JavierJia) [inline noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/TwitterFeedUsersTimelineDriver.java]: whenever you need the comments to describe the code, that is the time to create a separate function to wrap up the corresponding code. 

there should be a `getHistoricalUserData` function.

File: noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/TwitterFeedUsersTimelineDriver.java [hunk 2]
```diff
@@ -48,6 +60,46 @@ public void run(Config config) throws IOException {
         } catch (TwitterException te) {
             System.err.println("Twitter get user timeline Exception");
         }
+
+        //User Stream
```
Comment (JavierJia) [inline noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/TwitterFeedUsersTimelineDriver.java]: ditto

**Comment by JavierJia** [comment]
The code looks OK. However, I think the historical code and the stream data should not run together. We should get the historical data once, and run the stream code periodically in which case it doesn't need to get the historical data. 

Similar to what we want to achieve in the crawler project, we need a small trick to remember what is the last tweet id that we have crawled. Then it won't repeat its own history.

File: noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/TwitterFeedUsersTimelineDriver.java (hunk 1 above)
Comment (MoniMoledo) [inline noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/TwitterFeedUsersTimelineDriver.java]: Yes! Thank you!

**Comment by MoniMoledo** [comment]
I got your point. I will work to separate historical and stream data for now.
 After I can figure out a way to remember the last id to avoid repetition.

File: noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/CmdLineAux.java [hunk 3]
```diff
@@ -33,4 +35,17 @@ public static BufferedWriter createWriter(String fileName) throws IOException {
                 new OutputStreamWriter(zip, "UTF-8"));
         return bw;
     }
+
+    public static Long[] parseID(String[] stringIDs) throws CmdLineException{
```
Comment (JavierJia) [inline noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/CmdLineAux.java]: I found a better way here: http://stackoverflow.com/questions/754294/convert-an-array-of-primitive-longs-into-a-list-of-longs

File: noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/TwitterUserStreamDriver.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        BlockingQueue<String> queue = new LinkedBlockingQueue<String>(10000);
+        StatusesFilterEndpoint endpoint = new StatusesFilterEndpoint();
+        if (!usersID.isEmpty()) {
+            endpoint.followings(usersID);
+        }
+        Authentication auth = new OAuth1(config.getConsumerKey(), config.getConsumerSecret(), config.getToken(),
+                config.getTokenSecret());
+
+        // Create a new BasicClient. By default gzip is enabled.
+        twitterClient = new ClientBuilder()
```
Comment (JavierJia) [inline noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/TwitterUserStreamDriver.java]: It's usually better to merge the declaration and initialization together, like
`Client twitterClient = new ClientBuilder()`.

File: noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/TwitterUserStreamDriver.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+    public void run(Config config) throws IOException, CmdLineException {
+
+        List<Long> usersID = Arrays.asList(CmdLineAux.parseID(config.getTrackUsers()));
+
+        Client twitterClient;
+        boolean isConnected = false;
+        BlockingQueue<String> queue = new LinkedBlockingQueue<String>(10000);
+        StatusesFilterEndpoint endpoint = new StatusesFilterEndpoint();
+        if (!usersID.isEmpty()) {
```
Comment (JavierJia) [inline noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/TwitterUserStreamDriver.java]: will it work without any userIDs? If not we'd better to print the error and exit.

File: noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/TwitterUserStreamDriver.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        // Create a new BasicClient. By default gzip is enabled.
+        twitterClient = new ClientBuilder()
+                .hosts(Constants.STREAM_HOST)
+                .endpoint(endpoint)
+                .authentication(auth)
+                .processor(new StringDelimitedProcessor(queue))
+                .build();
+
+        // Establish a connection
+        BufferedWriter bw = CmdLineAux.createWriter("Tweet_UserStream_");
```
Comment (JavierJia) [inline noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/TwitterUserStreamDriver.java]: Choose a better file name?

Comment (JavierJia) [inline noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/TwitterUserStreamDriver.java]: Probably stick to the `gz` format as well? Then the downstream process could be unified.

**Comment by JavierJia** [comment]
looks good in general. just left some minor comments.

File: noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/TwitterUserStreamDriver.java (hunk 4 above)
Comment (MoniMoledo) [inline noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/TwitterUserStreamDriver.java]: Done

File: noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/CmdLineAux.java (hunk 3 above)
Comment (MoniMoledo) [inline noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/CmdLineAux.java]: Great! I was using the opposite method (toPrimitive()). However, this one fits better. Done

File: noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/TwitterUserStreamDriver.java (hunk 6 above)
Comment (MoniMoledo) [inline noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/TwitterUserStreamDriver.java]: @JavierJia 
I've changed the file name to "Tweet_Zika_Users_date.gz", any other suggestions? 

The format for this one is gz as well. I am using the same file creator for the 3 processes. The filename provided will be completed with the date and gz format.

File: noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/TwitterUserStreamDriver.java (hunk 5 above)
Comment (MoniMoledo) [inline noah/src/main/scala/edu/uci/ics/cloudberry/noah/feed/TwitterUserStreamDriver.java]: Done

**Comment by JavierJia** [comment]
👍

**Pull request merged**